File: node/fe/ui/pages/site.js
'use strict';

/**
 * Page object for the live storefront, automationexercise.com. Reads the
 * products grid, the search results, and the brand rail by their labels, so
 * the assertions are about the store, not the markup. A page that never
 * renders is ScreenNotReady, which the steps turn into Blocked -- a live site
 * being slow or down is not the site being wrong.
 */

const BASE = (process.env.SITE_URL || 'https://automationexercise.com').replace(/\/+$/, '');

class ScreenNotReady extends Error {}

class SitePage {
  constructor(page) { this.page = page; this.base = BASE; }

  async openProducts() {
    try {
      await this.page.goto(this.base + '/products', { waitUntil: 'domcontentloaded', timeout: 30_000 });
      await this.page.waitForSelector('.features_items .product-image-wrapper', { timeout: 25_000 });
    } catch (err) { throw new ScreenNotReady('products page did not render: ' + err.message); }
  }

  /** Product cards: name and price text, in grid order. */
  async products() {
    return this.page.$$eval('.features_items .product-image-wrapper .productinfo', (els) => els.map((el) => ({
      name: el.querySelector('p') ? el.querySelector('p').textContent.trim() : '',
      priceText: el.querySelector('h2') ? el.querySelector('h2').textContent.trim() : '',
    })));
  }

  async search(term) {
    try {
      await this.page.goto(this.base + '/products', { waitUntil: 'domcontentloaded', timeout: 30_000 });
      await this.page.fill('#search_product', term);
      await this.page.click('#submit_search');
      await this.page.waitForSelector('.features_items', { timeout: 25_000 });
    } catch (err) { throw new ScreenNotReady('search did not render: ' + err.message); }
    return this.products();
  }

  async openBrands() {
    try {
      await this.page.goto(this.base + '/products', { waitUntil: 'domcontentloaded', timeout: 30_000 });
      await this.page.waitForSelector('.brands_products .brands-name a', { timeout: 25_000 });
    } catch (err) { throw new ScreenNotReady('brands rail did not render: ' + err.message); }
    return this.page.$$eval('.brands_products .brands-name a', (els) => els.map((e) => e.textContent.replace(/\(\d+\)/, '').trim()).filter(Boolean));
  }
}

module.exports = { SitePage, ScreenNotReady, BASE };

File: python/fe/ui/pages/site.py
"""Page object for the live storefront, automationexercise.com. Mirror of
node/fe/ui/pages/site.js. Reads the grid, search and brand rail by label; a
page that never renders raises ScreenNotReady (grades Blocked)."""

from __future__ import annotations

import os

BASE = os.environ.get("SITE_URL", "https://automationexercise.com").rstrip("/")


class ScreenNotReady(Exception):
    pass


class SitePage:
    def __init__(self, page) -> None:
        self.page = page
        self.base = BASE

    def open_products(self):
        try:
            self.page.goto(self.base + "/products", wait_until="domcontentloaded", timeout=30000)
            self.page.wait_for_selector(".features_items .product-image-wrapper", timeout=25000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady(f"products page did not render: {err}") from err

    def products(self):
        return self.page.eval_on_selector_all(".features_items .product-image-wrapper .productinfo", """els => els.map(el => ({
            name: el.querySelector('p') ? el.querySelector('p').textContent.trim() : '',
            priceText: el.querySelector('h2') ? el.querySelector('h2').textContent.trim() : '',
        }))""")

    def search(self, term):
        try:
            self.page.goto(self.base + "/products", wait_until="domcontentloaded", timeout=30000)
            self.page.fill("#search_product", term)
            self.page.click("#submit_search")
            self.page.wait_for_selector(".features_items", timeout=25000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady(f"search did not render: {err}") from err
        return self.products()

    def open_brands(self):
        try:
            self.page.goto(self.base + "/products", wait_until="domcontentloaded", timeout=30000)
            self.page.wait_for_selector(".brands_products .brands-name a", timeout=25000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady(f"brands rail did not render: {err}") from err
        import re
        raw = self.page.eval_on_selector_all(".brands_products .brands-name a", "els => els.map(e => e.textContent)")
        return [b for b in (re.sub(r"\(\d+\)", "", t).strip() for t in raw) if b]
